### ISSUE TITLE
KMeans for Duplex

### ISSUE THREAD

**Issue by arieling** [dorarad/gansformer#25]
Thank you again for sharing the code!
If I understand correctly, when we use KMeans, k(Y) is deprecated. Another set of latent codes K is exploited to explicitly form spatial aware attention between image X and latent Y.  I have some questions with regards to intuition as well as implementation: 

1. In Sec 3.1.2. A notation of  K = a(Y, X) is used. I am super confused about this line.  

However, from line https://github.com/dorarad/gansformer/blob/main/training/network.py#L688. The Y is not used to compute centroid. Actually, if I understand correctly, the centroid is computed by K and the HxW image features. 

2. A very detailed question: why do you concat  _queries and queries - _queries in line?
https://github.com/dorarad/gansformer/blob/3c0bcdee049d82318cb2f9327c8d6c7808664cd2/training/network.py#L587

3. Seems positional encoding for latent is never used in k-means + simplex setting? The gradient only flows to g_mapping

Thank you again and looking forward to your reply!

**Comment by dorarad** [comment]
Hi, thanks so much for reaching out and apologies for the large delay in my response! 
1. Note that the centroids are computed by multiplying the `from_elements` which is indeed a function of the queries (and in turn of X) by the `from_to` tensor. When we use 0 iterations of K-means (in the case of simplex attention as in the pretrained models), then it is initialized to a [parameter](https://github.com/dorarad/gansformer/blob/3c0bcdee049d82318cb2f9327c8d6c7808664cd2/training/network.py#L614). However, when setting i > 0, the tensor becomes a [weighted averages](https://github.com/dorarad/gansformer/blob/3c0bcdee049d82318cb2f9327c8d6c7808664cd2/training/network.py#L708) over X elements based on the attention distribution between them and the Y latents, yielding the centroids.
Note that this is similar to how K-means operate, where the means are first initialized e.g. randomly, and then later are updated based on their interaction with the X elements. Here, each centroid is corresponding to a latent component from Y and its updates are based on the similarity between that Y latent and the X elements.

2. I added the _queries for backwards compatibility with the pre-trained models. It would be more elegant to just based them directly on queries. Will add a note about that!

3. hmm.. I'm actually surprised to hear that, since the code passed the latents' positional encodings to the transformer layer [here](https://github.com/dorarad/gansformer/blob/3c0bcdee049d82318cb2f9327c8d6c7808664cd2/training/network.py#L1290) and they are then should impact the `keys` through [this line](https://github.com/dorarad/gansformer/blob/3c0bcdee049d82318cb2f9327c8d6c7808664cd2/training/network.py#L684). I'll look into that further to see if there's anything incorrect in the implementation, but it should impact the attention distribution!
